File: JavaScript/1660-Correct a Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} from
 * @param {number} to
 * @return {TreeNode}
 */
var correctBinaryTree = function(root) {
    let queue = [root];
    let seen = new Set([root.val]);
    
    while (queue.length){
        let curr = queue.shift();
        if (curr.right){
            let grandRightNode = curr.right.right;
            if (grandRightNode && seen.has(grandRightNode.val) ){
                curr.right = null;
                return root;
            }
            seen.add(curr.right.val);
            queue.push(curr.right);
        }
        if (curr.left){
            let grandLeftNode = curr.left.right;
            if (curr.left.right && seen.has(grandLeftNode.val) ){
                curr.left = null;
                return root;
            }
            seen.add(curr.left.val);
            queue.push(curr.left);
        }
    }
};

File: C++/1660-Correct a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* dfs(TreeNode* node, unordered_set<TreeNode*>& visited) {
        if (!node) return nullptr;
        visited.insert(node);
        
        TreeNode* leftNode = nullptr;
        TreeNode* rightNode = nullptr;
        if (node->right) {
            if (visited.find(node->right) == visited.end()) rightNode = dfs(node->right, visited);
            else return nullptr;
        }
        if (node->left){
            leftNode = dfs(node->left, visited);
        }
        node->left = leftNode;
        node->right = rightNode;
        
        return node;
    }
    
    TreeNode* correctBinaryTree(TreeNode* root) {
        unordered_set<TreeNode*> visited;
        return dfs(root, visited);
    }
};